rfd/000?-database-access.md

```diff
@@ -0,0 +1,512 @@
+* AWS RDS for PostgreSQL.
+* PostgreSQL-compatible AWS Aurora.
+
+The following features are provided:
+
+* Connecting to the database through the Teleport proxy, incl. trusted
+  clusters support.
+* Limiting access to database instances by labels with Teleport roles.
+* Limiting access to individual databases (within a particular database
+  instance) and database users.
+* Auditing of database connections and executed queries.
+
+## Architecture
+
+Architecturally Database Access is very similar to [Application Access](0008-application-access.md).
+Teleport gains a new mode of operation, "database service", similar to "ssh
+service", "application service" or a "kube service".
+
+Database service runs inside the customer private network and proxies database
+client connections received from the Teleport proxy to the target database. It
+is also responsible for protocol parsing of supported databases and authorizing
+requests coming from the proxy (i.e. granting/denying access to particular
+databases/database users).
+
+```
+              |‾‾‾‾‾‾‾|  reverse  |‾‾‾‾‾‾‾‾‾‾‾‾|          |‾‾‾‾‾‾‾‾‾‾|
+psql -------> | proxy | <-------- | db service | -------> | postgres |
+              |_______|   tunnel  |____________|          |__________|
+```
+
+The database client (such as psql) will talk to Teleport web proxy port (`3080`
+by default) which will use multiplexing to detect the database protocol and
+dispatch to an appropriate service.
+
+## Authentication
+
+In this model, there are 3 points where authentication needs to happen.
+
+### Database client <---> proxy
+
+**Mutual TLS.** Database clients, such as psql, will use short-lived x509
+certificates issued by Teleport in order to authenticate with the proxy.
+
+The certificate metadata includes (via extensions) routing information about
+the target database which users log into using `tsh db login` command (see
+below for UX).
+
+### Proxy <---> database service
+
+**Mutual TLS.** The connection that is passed from proxy to a database service
+is upgraded to TLS as well in order to be able to pass identity information
+over the reverse tunnel.
+
+In order to do that, proxy will generate an ephemeral certificate signed
+by the user authority and re-encode all identity and routing information
+extracted from the certificate supplied by the client into it.
+
+### Database service <---> database instance
+
+**Mutual TLS (onprem).**  When connecting to an onprem database instance, the
+database service will use a client certificate for authentication, which it
+will generate on the fly according to the database requirements (for example,
+PostgreSQL requires the database user name to be encoded in the certificate
+as common name).
+
+This also means that the database needs to be configured with Teleport's
+certificate authority and (in general case) use server certificate issued by
+Teleport which can be done using `tctl auth sign` command (see below for UX).
```

Comment:
Can we also allow the user to give teleport a custom CA bundle to trust for db access?

Reply:
Yes, it's already supported, via `ca_path` (which I'll rename to `ca_cert_file` for consistency) field. It can be used both in case of rds/aurora to trust their root certificate, as well as providing ability for user to supply their own CA in case they're using their own cert/key pair. Although as I understand the trajectory Teleport is moving, this is not really a desirable scenario since we intend Teleport to be your certificate authority and issue all certificates (might be still helpful though for existing setups users may have).

Reply:
Gotcha.
The doc right now implies that `ca_path` is only supported for AWS.
Could you call this out explicitly somewhere?

Reply:
Yep, will do.